From ce42b8b62f65e2d9d05752f398fd027deda48ff2 Mon Sep 17 00:00:00 2001
From: "Philip, Avinash" <avinashphilip@ti.com>
Date: Mon, 2 Jul 2012 14:03:53 +0530
Subject: [PATCH 09/10] pwm: pwm-tiehrpwm: Add device-tree binding support
 EHRPWM driver

Adds device-tree binding support in EHRWPM driver and custom of_xlate
support. In custom of_xlate support, support for configuring polarity
also been provided. This will provide configuration of EHRPWM polarity
from client drivers device-tree.
Also size of pwm-cells set to 3 to support polarity configuration from
device tree.

Signed-off-by: Philip, Avinash <avinashphilip@ti.com>

Conflicts:
	drivers/pwm/pwm-tiehrpwm.c
---
 .../devicetree/bindings/pwm/ehrpwm-pwm.txt         |   24 ++++++++++++++
 drivers/pwm/pwm-tiehrpwm.c                         |   34 ++++++++++++++++++++
 2 files changed, 58 insertions(+), 0 deletions(-)
 create mode 100644 Documentation/devicetree/bindings/pwm/ehrpwm-pwm.txt

diff --git a/Documentation/devicetree/bindings/pwm/ehrpwm-pwm.txt b/Documentation/devicetree/bindings/pwm/ehrpwm-pwm.txt
new file mode 100644
index 0000000..a11543c
--- /dev/null
+++ b/Documentation/devicetree/bindings/pwm/ehrpwm-pwm.txt
@@ -0,0 +1,24 @@
+TI SOC EHRPWM based PWM controller
+
+Required properties:
+- compatible : Must be "ti, ehrpwm"
+- ti,hwmods : Must be "ehrpwm<n>", n being the instance number (0-based)
+- #pwm-cells: On EHRPWM the number of cells used to specify a PWM is 3. The
+  first cell specifies the per-chip index of the PWM to use, the second
+  cell is the period cycle in nanoseconds and the third cell is the
+  polarity of PWM output. Polarity 0 gives normal polarity and 1 gives
+  inversed polarity (inverse duty cycle)
+
+Note: Current implementation will fetch base address, irq and dma
+from omap hwmod data base during device registration.
+Future plan is to migrate hwmod data base contents into device tree
+blob so that, all the required data will be used from device tree dts
+file.
+
+Example:
+
+	ehrpwm0: ehrpwm@0 {
+		compatible = "ti, ehrpwm";
+		ti,hwmods = "ehrpwm0";
+		#pwm-cells = <3>;
+	};
diff --git a/drivers/pwm/pwm-tiehrpwm.c b/drivers/pwm/pwm-tiehrpwm.c
index 288f93e..4f1e467 100644
--- a/drivers/pwm/pwm-tiehrpwm.c
+++ b/drivers/pwm/pwm-tiehrpwm.c
@@ -108,6 +108,7 @@
 #define AQCSFRC_CSFA_DISSWFRC	(BIT(1) | BIT(0))
 
 #define NUM_PWM_CHANNEL		2	/* EHRPWM channels */
+#define PWM_CELL_SIZE		3
 
 struct ehrpwm_pwm_chip {
 	struct pwm_chip	chip;
@@ -393,6 +394,26 @@ static const struct pwm_ops ehrpwm_pwm_ops = {
 	.owner		= THIS_MODULE,
 };
 
+static struct pwm_device *of_ehrpwm_xlate(struct pwm_chip *chip,
+		const struct of_phandle_args *args)
+{
+	struct pwm_device *pwm;
+
+	if (chip->of_pwm_n_cells < PWM_CELL_SIZE)
+		return ERR_PTR(-EINVAL);
+
+	if (args->args[0] >= chip->npwm)
+		return ERR_PTR(-EINVAL);
+
+	pwm = pwm_request_from_chip(chip, args->args[0], NULL);
+	if (IS_ERR(pwm))
+		return pwm;
+
+	pwm_set_period(pwm, args->args[1]);
+	pwm_set_polarity(pwm, args->args[2]);
+	return pwm;
+}
+
 static int __devinit ehrpwm_pwm_probe(struct platform_device *pdev)
 {
 	int ret;
@@ -420,6 +441,8 @@ static int __devinit ehrpwm_pwm_probe(struct platform_device *pdev)
 
 	pc->chip.dev = &pdev->dev;
 	pc->chip.ops = &ehrpwm_pwm_ops;
+	pc->chip.of_xlate = of_ehrpwm_xlate;
+	pc->chip.of_pwm_n_cells = PWM_CELL_SIZE;
 	pc->chip.base = -1;
 	pc->chip.npwm = NUM_PWM_CHANNEL;
 
@@ -453,9 +476,20 @@ static int __devexit ehrpwm_pwm_remove(struct platform_device *pdev)
 	return pwmchip_remove(&pc->chip);
 }
 
+#if defined(CONFIG_OF)
+static const struct of_device_id omap_ehrpwm_of_match[] = {
+	{ .compatible = "ti,omap2-ehrpwm" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, omap_ehrpwm_of_match);
+#endif
+
 static struct platform_driver ehrpwm_pwm_driver = {
 	.driver = {
 		.name = "ehrpwm",
+#if defined(CONFIG_OF)
+		.of_match_table = of_match_ptr(omap_ehrpwm_of_match),
+#endif
 	},
 	.probe = ehrpwm_pwm_probe,
 	.remove = __devexit_p(ehrpwm_pwm_remove),
-- 
1.7.7.6

